fix: unify case-specific and global namelist args

haakon-e opened the pull request:
## Changes
Fix specification of both case-specific `namelist_args` and global (i.e. from `config["scm"]`) `namelist_args`.

Case-specific args has precedence over global args, which has precedence over default TC.jl args.

## Issue number (if applicable)
Generalizes a similar issue that was addressed in PR #394.
Closes #393.

## Checklist before requesting a review / merging
- [x] I have formatted the code using `julia --project=.dev .dev/climaformat.jl .`
- [x] I have updated all test manifests using `julia --project .dev/up_deps.jl .` with Julia 1.7.3.
- [x] If core features are added, I have added appropriate test coverage.
- [x] If new functions and structs are added, they are documented through docstrings.

haakon-e commented:
TODO (possibly in a future PR): Make the documentation of how the case-by-case & global `namelist_args` entries interact with each other and with default TC.jl arguments more accessible. Currently, I have only added basic function-documentation to `get_ref_model_kwargs`, which, although it clearly explains how the code works, is rather inaccessible for the average user who would only look at the `config` file. Perhaps we should add docs for common `config` file options, or similar?

ilopezgp commented on src/Pipeline.jl:
```diff
@@ -170,13 +169,35 @@ function init_calibration(config::Dict{Any, Any}; mode::String = "hpc", job_id::
+The namelist that defines a case is fetched from TC.jl for each case defined in `ref_config["case_name"]`.
+These can be overwritten in one of two ways;
+1. Define case-by-case overwrite entries in `ref_config["namelist_args"]`
+2. Define global overwrite entries with the keyword `global_namelist_args` (`Vector` of `Tuple`s).
```
```suggestion
2. Define global overwrite entries with the keyword argument `global_namelist_args` (`Vector` of `Tuple`s).
```

ilopezgp commented on test/ReferenceModels/runtests.jl:
```diff
@@ -1,6 +1,8 @@
 using CalibrateEDMF.ReferenceModels
 using CalibrateEDMF.TurbulenceConvectionUtils
+import CalibrateEDMF.HelperFuncs: get_entry
+import CalibrateEDMF.Pipeline: get_ref_model_kwargs
```
Maybe we can move `get_ref_model_kwargs` to the ReferenceModels module? This test suite should not depend on Pipeline.

ilopezgp reviewed (approved):
Great work! I think we can clean up a bit further (ModelEvaluators, etc), but this is a good step ahead.

LGTM after my suggested changes and merging commits.

haakon-e commented:
I went through the PR again, trying to fix the `ModelEvaluator` stuff you mentioned. I think I handled all of that, but a few points:
* I have removed `namelist_args` from the optional arguments of any function that relates to running TC.jl if the method signature requires either a `ReferenceModel` or a `ModelEvaluator`. This is to avoid any ambiguities/confusion about how `namelist_args` should be specified. That is, `namelist_args` should only be passed to the `ReferenceModel` constructor, and the `ReferenceModel` should contain precisely the namelist that is used in TC.jl, so no further overwrites are necessary. This includes: `run_SCM`, `eval_single_ref_model`, 
* and possibly somewhat controversially, `run_SCM_handler`. However, the variant of `run_SCM_handler` that acts directly on a `case` (not a `ReferenceModel`) keeps `namelist_args` in its signature since it's used in `grid_search`, among other places. 
* It appears that in the last PR (#382) I missed some references to namelists that are saved to the system. In particular, I have now deleted the method `namelist_directory` which appears to be not needed. Somewhat confusingly, this method is called in the `"Namelist modification"` `testset` in `test/TurbulenceConvectionUtils/runtests.jl`, which I thought should have errored out (since we don't write the namelist to file anymore), but either way ... I think a big chunk of that testset is obsolete, but would appreciate your thoughts on that, Ignacio.

ilopezgp commented:
> I went through the PR again, trying to fix the `ModelEvaluator` stuff you mentioned. I think I handled all of that, but a few points:
> 
> * I have removed `namelist_args` from the optional arguments of any function that relates to running TC.jl if the method signature requires either a `ReferenceModel` or a `ModelEvaluator`. This is to avoid any ambiguities/confusion about how `namelist_args` should be specified. That is, `namelist_args` should only be passed to the `ReferenceModel` constructor, and the `ReferenceModel` should contain precisely the namelist that is used in TC.jl, so no further overwrites are necessary. This includes: `run_SCM`, `eval_single_ref_model`,
> * and possibly somewhat controversially, `run_SCM_handler`. However, the variant of `run_SCM_handler` that acts directly on a `case` (not a `ReferenceModel`) keeps `namelist_args` in its signature since it's used in `grid_search`, among other places.
> * It appears that in the last PR ([refactor: house cleaning; remove scm_dir from ReferenceModel, rename functions, update docs #382](https://github.com/CliMA/CalibrateEDMF.jl/pull/382)) I missed some references to namelists that are saved to the system. In particular, I have now deleted the method `namelist_directory` which appears to be not needed. Somewhat confusingly, this method is called in the `"Namelist modification"` `testset` in `test/TurbulenceConvectionUtils/runtests.jl`, which I thought should have errored out (since we don't write the namelist to file anymore), but either way ... I think a big chunk of that testset is obsolete, but would appreciate your thoughts on that, Ignacio.

Overall the changes you propose here are great, restricting namelist_args to ReferenceModel construction time is what I envisioned when I opened #393 . I will take a look at the test set you mention.

ilopezgp commented on src/TurbulenceConvectionUtils.jl:
```diff
@@ -358,7 +341,7 @@ function run_SCM_handler(
     param_map::ParameterMap,
     namelist::Union{Dict, Nothing} = nothing,
-    namelist_args::Union{Vector, Nothing} = nothing,
+    namelist_args::Union{Nothing, Vector{<:Tuple}} = nothing,
```
I see what you mean by _controversial_ here, since we could in principle forget about this and add an upper level function that uses `m.namelist` and at the same time `!isnothing(namelist_args)`.

What do you think about adding something like this?

```
@assert (isnothing(namelist) ⊻ isnothing(namelist_args)) | 
        (!isnothing(namelist) && !isnothing(namelist_args)) "A non-default namelist and a namelist_arg cannot be passed at the same time to run_SCM_handler."
```

haakon-e replied:
I think this would be at odds with the `grid_search` code which currently uses both options. Specifically, it makes a call to `default_namelist` to ensure that a `seed` is not set.

https://github.com/CliMA/CalibrateEDMF.jl/blob/45c0373e56b239acdcd73918de27e6da9361d836/experiments/grid_search/grid_search.jl#L117-L129

Perhaps I don't quite understand the logic, is this the intended behavior?
```julia
julia> f(namelist, namelist_args) = (isnothing(namelist) ⊻ isnothing(namelist_args)) | 
               (!isnothing(namelist) && !isnothing(namelist_args))
f (generic function with 1 method)

julia> f(1,nothing)
true

julia> f(nothing,1)
true

julia> f(nothing,nothing)
false

julia> f(1,1)
true
```

haakon-e replied:
Perhaps we could create a separate helper-function for 
https://github.com/CliMA/CalibrateEDMF.jl/blob/a871046c041a9d8fd433c15488e7cdf70bb4c8fa/src/TurbulenceConvectionUtils.jl#L378-L382
and then remove the argument `namelist_args` from `run_SCM_handler` altogether?

haakon-e replied:
Opened an issue for this. See #407.

ilopezgp commented on test/Pipeline/runtests.jl:
```diff
@@ -24,14 +23,15 @@ config = get_config()
     ref_config["t_start"][1],
-    ref_config["t_end"][1],
+    ref_config["t_end"][1];
+    namelist_args = namelist_args,
```
if we are passing it this way, can you move the `namelist_args` to `config["reference"]`? The `config["scm"]["namelist_args"]` use should be limited to the `global_namelist_args` in the kwargs constructor in my opinion.

haakon-e replied:
I'll keep it in `config["scm"]` but remove `namelist_args` from the method signature here. I had just misinterpreted the test -- I thought it was supposed to have the `namelist_args` passed in there, but see now that it is the _true reference data_ that is constructed, not forward SCM evaluations.

Anyways my changing the signature also caused a test to fail and I now understand why. thanks!

ilopezgp commented on test/TurbulenceConvectionUtils/runtests.jl:
```diff
@@ -154,14 +153,20 @@ import CalibrateEDMF.HelperFuncs: do_nothing_param_map
 
-        run_scm_namelist_path = namelist_directory(res_dir, ref_models[1])
-        run_scm_namelist = JSON.parsefile(run_scm_namelist_path)
+        # HE: I think we should be able to delete everything below here, or what, Ignacio?
+        # this is technically related to the PR about namelists, so I'm not sure how this 
+        # test has passed since the namelist shouldn't exist at the path...
```
The run_reference_SCM still outputs a namelist, I think. I think we want to retain that behavior, we are just not using it in the Pipeline anymore.

ilopezgp commented on test/TurbulenceConvectionUtils/runtests.jl:
```diff
@@ -154,14 +153,20 @@ import CalibrateEDMF.HelperFuncs: do_nothing_param_map
-        run_scm_namelist = JSON.parsefile(run_scm_namelist_path)
+        # HE: I think we should be able to delete everything below here, or what, Ignacio?
+        # this is technically related to the PR about namelists, so I'm not sure how this 
+        # test has passed since the namelist shouldn't exist at the path...
+        run_scm_namelist = get_scm_namelist(ref_models[1])
         expected_run_scm_namelist = deepcopy(default_namelist)
-        expected_run_scm_namelist["turbulence"]["EDMF_PrognosticTKE"]["entrainment_factor"] = 0.15
-        expected_run_scm_namelist["turbulence"]["EDMF_PrognosticTKE"]["detrainment_factor"] = 0.52
+        expected_run_scm_namelist["turbulence"]["EDMF_PrognosticTKE"]["entrainment_factor"] = 0.13
+        expected_run_scm_namelist["turbulence"]["EDMF_PrognosticTKE"]["detrainment_factor"] = 0.51
```
This is a bit dangerous. We would not be testing anymore that setting new parameters for calibration actually sends those parameters to TurbulenceConvection. We need to test that TC was actually run with [0.15, 0.52] somehow.

haakon-e replied:
I missed the part of the test where those parameter values were passed by `u` and `u_names`. Reverted.

ilopezgp reviewed (approved):
LGTM, let's merge commits

ilopezgp commented:
bors try

haakon-e commented:
bors cancel

haakon-e commented:
bors r+

haakon-e commented:
bors r+

haakon-e commented:
bors r-

haakon-e commented:
bors r+
